src/main/java/ca/corefacility/bioinformatics/irida/repositories/filesystem/IridaFileStorageAzureUtilityImpl.java

```diff
@@ -0,0 +1,243 @@
+	}
+
+	/**
+	 * {@inheritDoc}
+	 */
+	@Override
+	public String getFileSize(Path file) {
+		String fileSize = "N/A";
+		try {
+			// We set the blobClient "path" to which we want to upload our file to
+			blobClient = containerClient.getBlobClient(getAzureFileAbsolutePath(file));
+			fileSize = FileUtils.humanReadableByteCount(blobClient.getProperties().getBlobSize(), true);
+		} catch (BlobStorageException e) {
+			logger.trace("Couldn't calculate size as the file was not found on azure [" + e + "]");
+		}
+		return fileSize;
+	}
+
+	/**
+	 * {@inheritDoc}
+	 */
+	@Override
+	public void writeFile(Path source, Path target, Path sequenceFileDir, Path sequenceFileDirWithRevision) {
+		// We set the blobClient "path" to which we want to upload our file to
+		blobClient = containerClient.getBlobClient(getAzureFileAbsolutePath(target));
+		try {
+			logger.trace("Uploading file to azure: [" + target.getFileName() + "]");
+			blobClient.uploadFromFile(source.toString(), false);
+			logger.trace("File uploaded to: [" + blobClient.getBlobUrl() + "]");
+		} catch (BlobStorageException e) {
+			logger.trace("Unable to upload file to azure [" + e + "]");
+		}
+	}
+
+	/**
+	 * {@inheritDoc}
+	 */
+	@Override
+	public boolean storageTypeIsLocal(){
+		return false;
+	}
+
+	/**
+	 * {@inheritDoc}
+	 */
+	public String getFileName(Path file) {
+		String fileName = "";
+		blobClient = containerClient.getBlobClient(getAzureFileAbsolutePath(file));
+		try {
+			// Since the file system is virtual the full file path is the file name.
+			// We split it on "/" and get the last token which is the actual file name.
+			String[] blobNameTokens = blobClient.getBlobName()
+					.split("/");
+			fileName = blobNameTokens[blobNameTokens.length - 1];
+		} catch (BlobStorageException e) {
+			logger.trace("Couldn't find file on azure [" + e + "]");
+		}
+
+		return fileName;
+	}
+
+	/**
+	 * {@inheritDoc}
+	 */
+	@Override
+	public boolean fileExists(Path file) {
+		blobClient = containerClient.getBlobClient(getAzureFileAbsolutePath(file));
+		if(blobClient.exists()) {
+			return true;
+		}
+		return false;
+	}
+
+	/**
+	 * {@inheritDoc}
+	 */
+	@Override
+	public InputStream getFileInputStream(Path file) {
+		try {
+			blobClient = containerClient.getBlobClient(getAzureFileAbsolutePath(file));
+		} catch (BlobStorageException e) {
+			logger.trace("Couldn't read file from azure [" + e + "]");
+		}
+		return blobClient.openInputStream();
+	}
+
+	/**
+	 * {@inheritDoc}
+	 */
+	@Override
+	public boolean isGzipped(Path file) throws IOException {
+		try (InputStream is = getFileInputStream(file)) {
+			byte[] bytes = new byte[2];
+			is.read(bytes);
+			return ((bytes[0] == (byte) (GZIPInputStream.GZIP_MAGIC))
+					&& (bytes[1] == (byte) (GZIPInputStream.GZIP_MAGIC >> 8)));
+		}
+	}
+
+	/**
+	 * Removes the leading "/" from the absolute path
+	 * returns the rest of the path.
+	 *
+	 * @param file
+	 * @return
+	 */
+	private String getAzureFileAbsolutePath(Path file) {
```

Comment:
Azure and AWS use a virtual filesystem. For example `/opt/irida/data/sequence/1/filename.fastq` would be the object. The file paths stored in the db have a leading slash which if we attempt to locate in the object store it tries to find containerName//object instead of containerName/object resulting in the bucket object (s3) or the blob(azure) not being found.

 I will update the comment to make this clear

Reply:
Ok.  Is there maybe a different implementation of `Path` that should be used for these cloud types?  Or maybe the `Path` class isn't the appropriate thing to store for this anymore?  I think `Path` is something that is assumed to be a file system reference.  Here since it's no longer a file on a file system we need to keep converting it.  Maybe we should be storing these paths in another way.  Lets talk about it next week.